buttercup/cogs/heatmap.py

```diff
@@ -0,0 +1,258 @@
+import io
+from typing import Dict, List, Optional
+
+import matplotlib.pyplot as plt
+from blossom_wrapper import BlossomAPI
+from discord import File
+from discord.ext.commands import Cog
+from discord_slash import SlashContext, cog_ext
+from discord_slash.utils.manage_commands import create_option
+from matplotlib.font_manager import FontProperties
+
+from buttercup.bot import ButtercupBot
+
+
+# Colors to use in the plots
+background_color = "#36393f"  # Discord background color
+text_color = "white"
+line_color = "white"
+intensity_colors = [
+    "#07102c",
+    "#0a1842",
+    "#0e2058",
+    "#11286e",
+    "#153184",
+    "#18399a",
+    "#1c41b0",
+    "#1f49c7",
+    "#2251dd",
+    "#3862e0",
+]
+
+
+Heatmap = Dict[int, Dict[int, int]]
+
+
+def init_heatmap_dict() -> Heatmap:
+    """Get the initial dictionary for the heatmap data.
+
+    For every day (1 = Monday, 7 = Sunday), it contains a dict entry
+    with an entry 0 for every hour.
+    """
+    heatmap_dict = {}
+
+    for day in range(1, 8):
+        day_dict = {}
+
+        for hour in range(0, 24):
+            day_dict[hour] = 0
+
+        heatmap_dict[day] = day_dict
+
+    return heatmap_dict
+
+
+def get_heatmap_dict(data: List[Dict[str, int]]) -> Heatmap:
+    """Populate the heatmap dictionary with the data returned by Blossom."""
+    heatmap_dict = init_heatmap_dict()
+
+    for entry in data:
+        day = entry["day"]
+        hour = entry["hour"]
+        count = entry["count"]
+
+        heatmap_dict[day][hour] += count
+
+    return heatmap_dict
+
+
+def get_cell_texts(heatmap_dict: Heatmap) -> List[List[str]]:
+    """Get the text for each cell of the table."""
+    rows = []
+
+    for day in heatmap_dict:
+        cols = []
+        day_data = heatmap_dict[day]
+
+        for hour in day_data:
+            hour_data = day_data[hour]
+            cell_text = "" if hour_data == 0 else str(hour_data)
+            cols.append(cell_text)
+
+        rows.append(cols)
+
+    return rows
+
+
+def get_max_value(heatmap_dict: Heatmap) -> int:
+    """Calculate the maximum value in the heatmap."""
+    max_value = 0
+
+    for day in heatmap_dict:
+        day_data = heatmap_dict[day]
+
+        for hour in day_data:
+            hour_data = day_data[hour]
+
+            max_value = max(max_value, hour_data)
+
+    return max_value
+
+
+def get_cell_colors(heatmap_dict: Heatmap) -> List[List[str]]:
+    """Get the color for each cell of the table."""
+    rows = []

```

Comment:
For the heatmap ourselves, I am not sure whether you are aware of Seaborn, which provides quite some niceheatmaps (see https://seaborn.pydata.org/generated/seaborn.heatmap.html). This also works great in tandem with pandas, as you might expect ;)

Reply:
Oh nice I'll check that out. In matplotlip it was a big pain to get right